File: src/java/BinarySearchTree.java

public class BinarySearchTree {
   private int size;
   private BSNode root;

   private class BSNode {
      public int key;
      public int val;
      public BSNode l;
      public BSNode r;

      public BSNode(int key, int val) {
         this.key = key;
         this.val = val;
         l = r = null;
      }
   }

   public void insert(int key, int val) {
      BSNode y = null, z = root;
      if (z == null) {
         root = new BSNode(key,val);
         size++;
         return;
      }
      while (z != null) {
         y = z;
         if (key == z.key) {
            z.val = val;
            return;
         }
         if (key < z.key) {
            z = z.l;
         }
         else z = z.r;
      }

      z = new BSNode(key,val);

      if (key < y.key) {
         y.l = z;
      }
      else y.r = z;

      size++;
   }

   public int get(int key) {
      BSNode z = root;
      while (z != null) {
         if (key == z.key) return z.val;
         else if (key < z.key) z = z.l;
         else z = z.r;
      }
      return -1;
   }

   private void transplant(BSNode u, BSNode up, BSNode v) {
      if (up == null) {
         root = v;
      }
      else if (u == up.l) up.l = v;
      else up.r = v;
   }

   public int delete(int key) {
      BSNode zp = null, z = root;
      while (z != null) {
         if (key == z.key) break;
         zp = z;
         if (key < z.key) z = z.l;
         else z = z.r;
      }
      if (z == null) return -1;

      BSNode sp = null, s;
      if (z.r != null && z.r.l == null) {
         s = z.r;
         transplant(z,zp,s);
         s.l = z.l;
      }
      else {
         s = z.l;
         if (s != null) while(s.r != null) { sp = s; s = s.r; }
         if (sp != null) sp.r = s.l;
         transplant(z,zp,s);
         if (s != null) {
            if (z.l != s) s.l = z.l;
            s.r = z.r;
         }
      }

      size--;
      return z.val;
   }

   private void inOrderPrintHelper(BSNode n) {
      if (n == null) return;
      System.out.print("l");
      inOrderPrintHelper(n.l);
      System.out.printf("%d ",n.key);
      System.out.print("r");
      inOrderPrintHelper(n.r);
   }

   public void inOrderPrint() {
      inOrderPrintHelper(root);
      System.out.println();
   }
}

File: src/c/binary_search_tree.c

#include "binary_search_tree.h"
#include <stdlib.h>
#include <stdio.h>

void init_bs(struct bs_tree * t) {
   t->size = 0;
   t->root = 0;
}

void insert_bs(int key, int val, struct bs_tree * t) {
   struct bs_node * y, * z = t->root;
   if (!z) {
      t->root = malloc(sizeof(struct bs_node));
      t->root->key = key; t->root->val = val;
      t->root->l = t->root->r = 0;
      t->size++;
      return;
   }
   while (z) {
      y = z;
      if (key == z->key) {
         z->val = val;
         return;
      }
      if (key < z->key) {
         z = z->l;
      }
      else z = z->r;
   }

   z = malloc(sizeof(struct bs_node));
   z->key = key; z->val = val;
   z->l = 0; z->r = 0;

   if (key < y->key) {
      y->l = z;
   }
   else y->r = z;

   t->size++;
}

int get_bs(int key, struct bs_tree * t) {
   struct bs_node * z = t->root;
   while (z) {
      if (key == z->key) return z->val;
      else if (key < z->key) z = z->l;
      else z = z->r;
   }
   return -1;
}

void transplant_bs(struct bs_node * u, struct bs_node * up, struct bs_node * v, struct bs_tree * t) {
   if (!up) {
      t->root = v;
   }
   else if (u == up->l) up->l = v;
   else up->r = v;
}

int delete_bs(int key, struct bs_tree * t) {
   struct bs_node * zp = 0, * z = t->root;
   while (z) {
      if (key == z->key) break;
      zp = z;
      if (key < z->key) z = z->l;
      else z = z->r;
   }
   if (!z) return -1;

   struct bs_node * sp = 0, * s;
   if (z->r && !z->r->l) {
      s = z->r;
      transplant_bs(z,zp,s,t);
      s->l = z->l;
   }
   else { 
      s = z->l;
      if (s) while(s->r) { sp = s; s = s->r; }
      if (sp) sp->r = s->l;
      transplant_bs(z,zp,s,t);
      if (s) {
         if (z->l != s) s->l = z->l;
         s->r = z->r;
      }
   }

   int ret = z->val;
   free(z);
   t->size--;
   return ret;
}

void destroy_helper_bs(struct bs_node * n) {
   if (n == 0) return;
   destroy_helper_bs(n->l);
   destroy_helper_bs(n->r);
   free(n);
}

void in_order_print_helper_bs(struct bs_node * n) {
   if (n == 0) return;
   printf("l");
   in_order_print_helper_bs(n->l);
   printf("%d ",n->key);
   printf("r");
   in_order_print_helper_bs(n->r);
}

void destroy_bs(struct bs_tree * t) {
   destroy_helper_bs(t->root);
   free(t);
}

void in_order_print_bs(struct bs_tree * t) {
   in_order_print_helper_bs(t->root);
   printf("\n");
}
